Expose a test fixture helper class

chapulina opened the pull request:
# 🎉 New feature

## Summary
<!--Explain changes made, the expected behavior, and provide any other additional
context (e.g., screenshots, gifs) if appropriate.-->

It's not always straightforward to write automated tests that leverage `ign-gazebo` from external projects. Internal to Gazebo, we have various helper classes in [test/helpers](https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo5/test/helpers) that make it easier to write integration tests. The goal of this PR is to expose a public API that makes it easier to write tests similar to those.

The main class being added in this PR is the `TestFixture`. See the added example for how to use it in an external project together with GTest. Example usage from the classes's docs:

```
/// // Load a world with a fixture
/// ignition::gazebo::TestFixture fixture("path_to.sdf");
///
/// // Register callbacks, for example:
/// fixture.OnPostUpdate([&](const gazebo::UpdateInfo &,
///   const gazebo::EntityComponentManager &_ecm)
///   {
///     // Add expectations here
///   }
///
/// // Run the server
/// fixture.Server()->Run(true, 1000, false);
```

Besides `test/helpers`, a lot of the inspiration here is coming from @arjo129 's work on a separate project.

### Differences from test/helpers

While working on the public API I made some improvements over what we are currently using internally:

* Added new `Server::AddSystem` APIs that allow us to load a plugin created on the fly, instead of needing to load a shared library. (i.e. no more need for `libMockPlugin.so`, see the `HelperSystem` instead)
* The `TestFixture`, unlike `test/helpers/Relay`, instantiates a server automatically. This saves even more boilerplate in test code.

## Test it
<!--Explain how reviewers can test this new feature manually.-->

Run the new standalone example. I made an effort to use classes like `World` and `Model` on the example test in order to make it as approachable as possible. (i.e. no `Each` calls)

## TODO

TODO before merging:

- [x] Add tests for `Server::AddSystem`
- [x] Add tests for `TestFixture`
- [x] I'd like to add a 2nd example test that is more realistic, sending commands and reading sensor data
- [x] General cleanup (docs, style, etc)

TODO in the future:

* ~~Tying into the simulation update loop is very powerful, but it can be inconvenient for doing one-time setup operations, such as getting a model's entity ID.~~ Done in #930 

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

ahcorde commented on examples/standalone/custom_server/CMakeLists.txt:
```diff
@@ -2,10 +2,10 @@ cmake_minimum_required(VERSION 3.10.2 FATAL_ERROR)
 
```
This is not part of this PR. Do you mind to remove it ?

chapulina replied:
I noticed it while working on this PR. It could go on its own PR, but I thought it was small enough to put it here, this way we save some CI cycles

ahcorde commented on examples/standalone/gtest_setup/gravity.sdf:
```diff
@@ -0,0 +1,33 @@
+            </sphere>
+          </geometry>
+        </collision>
+
```
minor style issue
```suggestion
```

ahcorde commented on include/ignition/gazebo/TestFixture.hh:
```diff
@@ -0,0 +1,89 @@
+{
+  /// \brief Constructor
+  /// \param[in] _path Path to SDF file.
+  public: TestFixture(const std::string &_sdf);
```
```suggestion
  public: TestFixture(const std::string &_path);
```

ahcorde commented on src/TestFixture.cc:
```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
```
```suggestion
 * Copyright (C) 2021 Open Source Robotics Foundation
```

ahcorde commented on src/TestFixture.cc:
```diff
@@ -0,0 +1,155 @@
+}
+
+//////////////////////////////////////////////////
+TestFixture::TestFixture(const std::string &_sdf)
```
```suggestion
TestFixture::TestFixture(const std::string &_path)
```

arjo129 reviewed (commented):
Regarding the TODO I've opened #930. This will make it easier to perform tests at startup. Most of my comments are minor nitpicks.

chapulina commented:
This PR is ready for review. I recommend starting by reviewing https://github.com/ignitionrobotics/ign-gazebo/pull/935 and https://github.com/ignitionrobotics/ign-gazebo/pull/936 first, which are subsets of this PR and will make it shorter.

Once this is approved targeted at Fortress, I'm planning to retarget to Citadel, update as needed, merge, and forward-port up to Fortress.

chapulina commented:
@mjcarroll , I merged from `main` and cleaned it up. It's ready for review :+1:

caguero reviewed (approved):
It looks good to me. The only comment I have is the example requires CMake 3.11 and Ubuntu Bionic ships with 3.10. We claim that Fortress is supported in Ubuntu Bionic. I think this is OK as it's one example and there are other alternative ways to install CMake.

chapulina commented:
The PR was approved at 9963485 when targeting `main` (v6). I rebased onto `ign-gazebo3` and updated the API to work there on a05623d.

I'm going to merge into `ign-gazebo3` when CI comes back happy.

If anyone wants to use this with `main` before it's forward-ported, checkout 9963485.
